test_util/provide_aws.py

```diff
@@ -0,0 +1,152 @@
+        """
+        logging.info('Requesting AWS CloudFormation...')
+        cf_parameters = []
+        for k, v in user_parameters.items():
+            cf_parameters.append({'ParameterKey': k, 'ParameterValue': v})
+        return self.resource('cloudformation').create_stack(
+                StackName=name,
+                TemplateURL=template_url,
+                DisableRollback=True,
+                TimeoutInMinutes=AWS_WAITING_TIMEOUT,
+                Capabilities=['CAPABILITY_IAM'],
+                Parameters=cf_parameters)
+
+
+class CFStack():
+    def __init__(self, stack_name, boto_wrapper):
+        self.stack_name = stack_name
+        self.boto_wrapper = boto_wrapper
+
+    def attach(self):
+        self.stack = self.boto_wrapper.resource('cloudformation').Stack(self.stack_name)
+
+    def wait_for_stack_creation(self):
+        self.wait_for_transition('CREATE_IN_PROGRESS', 'CREATE_COMPLETE')
+
+    def wait_for_stack_deletion(self):
+        self.wait_for_transition('DELETE_IN_PROGRESS', 'DELETE_COMPLETE')
+
+    def wait_for_transition(self, start_status, end_status):
+        final_status, status_reason = self.wait_for_status_change(start_status)
+        assert end_status == final_status, "Expected status {} but found {}. Reason: {}".format(
+                end_status, final_status, status_reason)
+
+    @retrying.retry(wait_fixed=5000, stop_max_delay=AWS_WAITING_TIMEOUT*60*1000,
+                    retry_on_result=lambda res: res is False,
+                    retry_on_exception=lambda ex: False)
+    def wait_for_status_change(self, status):
+        """NOTE: need to get stack_id before looping on wait or else method will
+        raise an excetion if the stack deletes and it cannot find stack ID
+        """
+        logging.info('Waiting for stack to finish: '+status)
+        current_status, current_reason = self.get_stack_status()
+        if(current_status == status):
+            return False
+        return current_status, current_reason
+
+    def get_stack_status(self):
+        """NOTE: stack_id must be externally defined for deleted stacks
+        """
+        stack_info = self.get_stack_info()
+        if 'StackStatusReason' in stack_info:
+            return stack_info['StackStatus'], stack_info['StackStatusReason']
+        else:
+            return stack_info['StackStatus'], 'None'
+
+    def get_stack_info(self):
+        """NOTE: stack_id must be externally defined for deleted stacks
+        """
+        stack_info = self.boto_wrapper.client('cloudformation').describe_stacks(StackName=self.stack.stack_id)
+        logging.debug('Stack Description for {}: {}'.format(self.stack.stack_id, stack_info))
+        return stack_info['Stacks'][0]
+
+
+class DcosCfSimple(CFStack):
+    @classmethod
+    def create(cls, stack_name, template_url, public_agents, private_agents,
+               admin_location, key_pair_name, boto_wrapper):
+        ret = DcosCfSimple(stack_name, boto_wrapper)
```

Comment:
Also saves us from having to do all the late initialization of self.stack, as well as the slightly awkward initialization of the CFStack internal structure which really breaks separation of concerns.